### PR TITLE
b/170423927: #3 replace snakeName with jsonName using syntax parsing

### PR DESCRIPTION
**Pull request by TAOXUY**
**Description**:  This PR will do syntax parsing the uri template and rebuilding it along with replacement of snakeName for jsonName, which makes more theoretically correct.

**Context**: The issue has been described in #371 and in that PR we workaround it by regex replacement. 

**Note**:  `{var}`, `{var=*}` are equal in syntax and which type expression is used is dropped during parsing and no points to stick on it. `ReplaceVariableFieldName` will all use `{var=*}`, which is also consistent with other syntaxes `{var=a/b/*}` or `{var=**}`.

**Details**:
- expose `ReplaceVariableFieldInUriTemplate` to realize this functionality and add its tests
- use it in service_info.go and remove the regex replacement related code

### CODE REVIEW

File: src/go/util/httppattern/uri_template_impl.go [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	inVariable bool
+	segments   []string
+	verb       string
+	variables  []*variable
+}
+
+// `serializeUriTemplate` serialize the UriTemplate object into the string representation.
+// It assumes the input is valid uriTemplate generated from `Parse()` and it
+// won't do any validation.
+func serializeUriTemplate(input *UriTemplate, fieldNameMapping map[string]string) string {
```
Comment (qiwzhang) [inline src/go/util/httppattern/uri_template_impl.go:61]: This is not just serialize?  It is serialize + replacement.   I am wondering if it is possible 
1) doing the replacement in UrlTemplate
2) then serailize() just do a searialize.

I think the code will be cleaner

File: src/go/util/httppattern/uri_template_impl.go [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+	input      string
+	tb         int
+	te         int
+	inVariable bool
+	segments   []string
+	verb       string
+	variables  []*variable
+}
+
+// `serializeUriTemplate` serialize the UriTemplate object into the string representation.
```
Comment (qiwzhang) [inline src/go/util/httppattern/uri_template_impl.go]: Also can we move serialize and replacement code into a separate file.  Not to mix with url_template parsing code.

File: src/go/configgenerator/listener_generator_test.go [hunk 3]
```diff
@@ -962,13 +962,13 @@ func TestPathMatcherFilter(t *testing.T) {
 						{
 							Selector: "1.cloudesf_testing_cloud_goog.Foo",
 							Pattern: &annotationspb.HttpRule_Get{
-								Get: "foo/{foo_bar}",
```
Comment (nareddyt) [inline src/go/configgenerator/listener_generator_test.go:965]: Is this a breaking change? We allowed HttpRule without a prefix `/`, but now we enforce it?

File: src/go/configinfo/service_info.go [hunk 4]
```diff
@@ -724,19 +724,28 @@ func (s *ServiceInfo) processTypes() error {
 			}
 		}
 
+		snakeNameToJsonNameForUriTemplates := func(m *MethodInfo, snakeNameToJsonName map[string]string) error {
+			for _, httpRule := range m.HttpRule {
+				if uriTemplate, err := httppattern.ReplaceVariableFieldName(httpRule.UriTemplate, snakeNameToJsonName); err != nil {
+					return fmt.Errorf("fail to replace snake name with json name in uri template, %v", err)
```
Comment (nareddyt) [inline src/go/configinfo/service_info.go:730]: Include the the original template or operation name in the error

File: src/go/util/httppattern/uri_template_impl.go [hunk 5]
```diff
@@ -0,0 +1,454 @@
+// Copyright 2019 Google LLC
```
Comment (nareddyt) [inline src/go/util/httppattern/uri_template_impl.go:1]: Splitting into impl is c++ style, not golang style. We don't split the public interface from the implementation like this. How about change this file as `uri_template_parser.go`, and then move the 'serialize`  to `http_template.go`?

File: src/go/util/httppattern/uri_template_impl.go (hunk 1 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template_impl.go]: Done.

File: src/go/util/httppattern/uri_template_impl.go (hunk 2 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template_impl.go]: Keep parsing code into uri_template_parser.go

File: src/go/configgenerator/listener_generator_test.go (hunk 3 above)
Comment (TAOXUY) [reply src/go/configgenerator/listener_generator_test.go:965]: Starting with '/' is required. 

https://cloud.google.com/endpoints/docs/grpc-service-config/reference/rpc/google.api#path-template-syntax

File: src/go/util/httppattern/uri_template.go [hunk 6]
```diff
@@ -45,365 +47,100 @@ type variable struct {
...
-		Verb:      p.verb,
-		Variables: p.variables,
-	}
+	return uriTemplate.String(), nil
 }
 
-func (p *parser) parse() bool {
-	if !p.parseTemplate() || !p.consumeAllInput() {
-		return false
+func (u *UriTemplate) String() string {
```
Comment (qiwzhang) [inline src/go/util/httppattern/uri_template.go:64]: How about ToString()?

File: src/go/util/httppattern/uri_template_test.go [hunk 7]
```diff
@@ -1000,10 +999,136 @@ func TestUriTemplateParseError(t *testing.T) {
...
-			if Parse(tc) != nil {
+			if ParseUriTemplate(tc) != nil {
 				t.Fatalf("succeed parsing %s but expect to fail", tc)
 			}
 		})
 
 	}
 }
+
+func TestReplaceVariableFieldInUriTemplateRebuild(t *testing.T) {
```
Comment (qiwzhang) [inline src/go/util/httppattern/uri_template_test.go:1010]: How about split into two test files,  one for url_template and another for parer.

File: src/go/util/httppattern/uri_template.go (hunk 6 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template.go:64]: `String()` is the the `ToString()` in [golang](https://stackoverflow.com/questions/41232904/is-there-a-go-equivalent-of-tostring-that-fmt-print-will-use).

File: src/go/util/httppattern/uri_template_test.go (hunk 7 above)
Comment (TAOXUY) [reply src/go/util/httppattern/uri_template_test.go:22]: Done.

**Comment by TAOXUY** [comment]
PTAL

**Review by qiwzhang — APPROVED** [review]

**Review by nareddyt — APPROVED** [review]

**Pull request merged**